addons/l10n_jo_edi/models/account_move.py

```diff
@@ -0,0 +1,159 @@
+        return super().action_post()
+
+    def _l10n_jo_build_jofotara_headers(self):
+        self.ensure_one()
+        return {
+            'Client-Id': self.company_id.l10n_jo_edi_client_id,
+            'Secret-Key': self.company_id.l10n_jo_edi_secret_key,
+            'Content-Type': 'application/json',
+        }
+
+    def _submit_to_jofotara(self):
+        self.ensure_one()
+        headers = self._l10n_jo_build_jofotara_headers()
+        params = '{"invoice": "%s"}' % self.l10n_jo_edi_xml_attachment_id.datas.decode('ascii')
+
+        try:
+            response = requests.post(JOFOTARA_URL, data=str(params), headers=headers, timeout=50, verify=False)
+        except requests.exceptions.Timeout:
+            return "Request time out! Please try again."
+        except (requests.exceptions.RequestException, requests.exceptions.HTTPError) as e:
+            return f"Invalid request: {e}"
+
+        response_text = response.content.decode()
+        if not response.ok:
+            return f"Request failed: {response_text}"
+        dict_response = json.loads(response_text)
+        self.l10n_jo_edi_qr = str(dict_response.get('EINV_QR', ''))
+
+    def _l10n_jo_qr_code_src(self):
+        self.ensure_one()
+        return f'/report/barcode/?barcode_type=QR&value={self.l10n_jo_edi_qr}&width=200&height=200'
+
+    def _l10n_jo_edi_get_xml_attachment_name(self):
+        return f"{self.name.replace('/', '_')}_edi.xml"
+
+    def _l10n_jo_validate_config(self):
+        error_msg = ''
+        if not self.company_id.l10n_jo_edi_client_id:
+            error_msg += "Client ID is missing.\n"
+        if not self.company_id.l10n_jo_edi_secret_key:
+            error_msg += "Secret key is missing.\n"
+        if not self.company_id.l10n_jo_edi_taxpayer_type:
+            error_msg += "Taxpayer type is missing.\n"
+        if not self.company_id.l10n_jo_edi_sequence_income_source:
+            error_msg += "Activity number (Sequence of income source) is missing.\n"
+
+        if error_msg:
+            raise UserError(f"{error_msg} To set: Configuration > Settings > Electronic Invoicing (Jordan)")
+
+    @api.depends("l10n_jo_edi_is_needed")
+    def _compute_l10n_jo_edi_uuid(self):
+        for invoice in self:
+            if invoice.l10n_jo_edi_is_needed and not invoice.l10n_jo_edi_uuid:
+                invoice.l10n_jo_edi_uuid = uuid.uuid4()
+
+    def _l10n_jo_edi_send(self):
+        self._l10n_jo_validate_config()
+        for invoice in self:
+            invoice_xml = self.env["ir.attachment"].create(
+                {
+                    "res_model": "account.move",
+                    "res_id": invoice.id,
+                    "res_field": "l10n_jo_edi_xml_attachment_file",
+                    "name": invoice._l10n_jo_edi_get_xml_attachment_name(),
+                    "raw": invoice.env['account.edi.xml.ubl_21.jo']._export_invoice(invoice),
+                }
+            )
+            invoice.invalidate_recordset(
```

Comment:
still needed?

Reply:
Why not? Also, the locking part is not so clear for me.